File: backend/apps/usuarios/services/reserva_service.py
from datetime import datetime

from django.utils import timezone

from ..models import Reserva, Notificacion
from ..data.estados_reserva import EstadoReserva
from ..data.tipos_notificacion import TipoNotificacion
from ..repositories.reserva_repository import ReservaRepository
from ..repositories.horario_repository import HorarioRepository
from ..repositories.notificacion_repository import NotificacionRepository
from .exceptions import ReglaNegocioError


class ReservaService:

    @staticmethod
    def crear(serializer):
        """El serializer ya validó las reglas de creación
        (horario disponible, sin duplicados, etc). Aquí solo
        se guarda y se dispara el efecto secundario: notificar."""
        reserva = serializer.save()

        NotificacionRepository.crear(
            usuario=reserva.profesional.usuario,
            reserva=reserva,
            tipo=TipoNotificacion.NUEVA_RESERVA,
            mensaje=(
                f"Nueva reserva de {reserva.cliente.username} "
                f"para el {reserva.fecha} a las {reserva.hora}."
            )
        )

        return reserva

    @staticmethod
    def confirmar(reserva_id):
        reserva = ReservaRepository.buscar_por_id(reserva_id)

        if reserva.estado != EstadoReserva.PENDIENTE:
            raise ReglaNegocioError(
                "Solo se pueden confirmar reservas pendientes."
            )

        reserva.estado = EstadoReserva.CONFIRMADA
        reserva.save()

        NotificacionRepository.crear(
            usuario=reserva.cliente,
            reserva=reserva,
            tipo=TipoNotificacion.CONFIRMACION,
            mensaje=(
                f"Tu reserva para {reserva.servicio.nombre} "
                f"el {reserva.fecha} a las {reserva.hora} "
                f"ha sido confirmada."
            )
        )

        return reserva

    @staticmethod
    def cancelar(reserva_id):
        reserva = ReservaRepository.buscar_por_id(reserva_id)

        if reserva.estado in [
            EstadoReserva.CANCELADA,
            EstadoReserva.COMPLETADA,
        ]:
            raise ReglaNegocioError(
                "Esta reserva no se puede cancelar."
            )

        reserva.estado = EstadoReserva.CANCELADA
        reserva.save()

        NotificacionRepository.crear(
            usuario=reserva.cliente,
            reserva=reserva,
            tipo=TipoNotificacion.CANCELACION,
            mensaje=(
                f"Tu reserva para {reserva.servicio.nombre} "
                f"el {reserva.fecha} a las {reserva.hora} "
                f"ha sido cancelada."
            )
        )

        return reserva

    @staticmethod
    def rechazar(reserva_id):
        reserva = ReservaRepository.buscar_por_id(reserva_id)

        if reserva.estado != EstadoReserva.PENDIENTE:
            raise ReglaNegocioError(
                "Solo se pueden rechazar reservas pendientes."
            )

        reserva.estado = EstadoReserva.RECHAZADA
        reserva.save()

        return reserva

    @staticmethod
    def reprogramar(reserva_id, nueva_fecha_str, nueva_hora_str):
        reserva = ReservaRepository.buscar_por_id(reserva_id)

        if reserva.estado == EstadoReserva.CANCELADA:
            raise ReglaNegocioError(
                "No se puede reprogramar una reserva cancelada."
            )

        if reserva.estado == EstadoReserva.COMPLETADA:
            raise ReglaNegocioError(
                "No se puede reprogramar una reserva completada."
            )

        if not nueva_fecha_str or not nueva_hora_str:
            raise ReglaNegocioError(
                "Debes enviar una nueva fecha y una nueva hora."
            )

        try:
            fecha = datetime.strptime(nueva_fecha_str, "%Y-%m-%d").date()
            hora = datetime.strptime(nueva_hora_str, "%H:%M:%S").time()
        except ValueError:
            raise ReglaNegocioError(
                "La fecha debe ser YYYY-MM-DD y la hora HH:MM:SS."
            )

        if fecha < timezone.localdate():
            raise ReglaNegocioError(
                "No puedes reprogramar para una fecha pasada."
            )

        profesional = reserva.profesional
        horarios = HorarioRepository.activos_por_dia(
            profesional.id, fecha.weekday()
        )

        if not horarios.exists():
            raise ReglaNegocioError(
                "El profesional no tiene horario disponible para ese día."
            )

        horario_valido = HorarioRepository.existe_horario_valido(
            profesional.id, fecha.weekday(), hora
        )

        if not horario_valido:
            raise ReglaNegocioError(
                "La nueva hora está fuera del horario de atención."
            )

        if ReservaRepository.existe_conflicto(
            profesional.id, fecha, hora, excluir_id=reserva.id
        ):
            raise ReglaNegocioError(
                "El nuevo horario ya está reservado."
            )

        reserva.fecha = fecha
        reserva.hora = hora
        reserva.estado = EstadoReserva.REPROGRAMADA
        reserva.save()

        NotificacionRepository.crear(
            usuario=reserva.cliente,
            reserva=reserva,
            tipo=TipoNotificacion.REPROGRAMACION,
            mensaje=(
                f"Tu reserva para {reserva.servicio.nombre} "
                f"ha sido reprogramada para el {reserva.fecha} "
                f"a las {reserva.hora}."
            )
        )

        return reserva

    @staticmethod
    def disponibilidad(profesional_id, fecha_str):
        if not profesional_id or not fecha_str:
            raise ReglaNegocioError(
                "Debes enviar profesional y fecha."
            )

        try:
            fecha = datetime.strptime(fecha_str, "%Y-%m-%d").date()
        except ValueError:
            raise ReglaNegocioError(
                "La fecha debe tener el formato YYYY-MM-DD."
            )

        horarios = HorarioRepository.activos_por_dia(
            profesional_id, fecha.weekday()
        )

        reservas = ReservaRepository.del_dia_excluyendo_canceladas(
            profesional_id, fecha
        )

        horas_ocupadas = [
            reserva.hora.strftime("%H:%M:%S")
            for reserva in reservas
        ]

        return {
            "profesional": profesional_id,
            "fecha": str(fecha),
            "horarios": [
                {
                    "hora_inicio": horario.hora_inicio,
                    "hora_fin": horario.hora_fin
                }
                for horario in horarios
            ],
            "horas_ocupadas": horas_ocupadas
        }

File: backend/src/services/reserva.service.ts
import { ReservaRepository } from "../repositories/reserva.repository";
import { MailService, DatosReservaCorreo } from "./mail.service"; // Ajusta la ruta si es necesario
import { ESTADOS_RESERVA } from "../data/estadosReserva";

export class ReservaService {
  /**
   * Crea una nueva reserva y envía correo de confirmación al cliente.
   */
  static async crear(datos: any, user: any) {
    // 1. Validar si ya existe un conflicto de horario para el profesional
    const existeConflicto = await ReservaRepository.existeConflicto(
      datos.profesional,
      datos.fecha,
      datos.hora
    );

    if (existeConflicto) {
      throw new Error("El profesional ya tiene una reserva en la fecha y hora seleccionadas.");
    }

    // 2. Asignar el cliente desde el usuario autenticado si no viene en el body
    const payloadReserva = {
      ...datos,
      cliente: datos.cliente || user?.id || user?._id,
    };

    // 3. Crear el registro base
    const nuevaReserva = await ReservaRepository.crear(payloadReserva);

    // 4. Buscar la reserva recien creada populando relaciones (cliente, servicio, profesional)
    const reservaPopulada: any = await ReservaRepository.buscarPorId(
      String(nuevaReserva._id)
    );

    // 5. Armar el payload para el correo
    const datosCorreo: DatosReservaCorreo = {
      to: user?.email || reservaPopulada?.cliente?.email,
      nombreCliente: user?.nombre || reservaPopulada?.cliente?.nombre || "Cliente",
      servicio: reservaPopulada?.servicio?.nombre || "Servicio contratado",
      profesional: reservaPopulada?.profesional?.nombre,
      fecha: reservaPopulada?.fecha || datos.fecha,
      hora: reservaPopulada?.hora || datos.hora,
      codigo: String(nuevaReserva._id).slice(-6).toUpperCase(),
    };

    // 6. Enviar notificación por correo de forma asíncrona (sin bloquear la respuesta HTTP)
    MailService.enviarConfirmacionReserva(datosCorreo).catch((error) => {
      console.error("[MAIL ERROR] Error al enviar correo de creación de reserva:", error);
    });

    return reservaPopulada || nuevaReserva;
  }

  /**
   * Obtiene la lista de reservas asociadas al usuario autenticado.
   */
  static async listar(user: any) {
    // Si el usuario es un cliente, se pueden filtrar sus reservas,
    // o bien listar todas si es administrador/profesional.
    if (user?.rol === "CLIENTE") {
      // Ajustar según el esquema de roles de tu app
      return await ReservaRepository.deClienteConProfesional(user.id, user.profesionalId);
    }
    return await ReservaRepository.listarTodas();
  }

  /**
   * Busca el detalle de una reserva por ID.
   */
  static async buscar(id: string) {
    const reserva = await ReservaRepository.buscarPorId(id);
    if (!reserva) {
      throw new Error("Reserva no encontrada");
    }
    return reserva;
  }

  /**
   * Confirma una reserva y notifica al cliente por correo.
   */
  static async confirmar(id: string, user: any) {
    const reserva: any = await ReservaRepository.buscarPorId(id);
    if (!reserva) {
      throw new Error("Reserva no encontrada");
    }

    reserva.estado = ESTADOS_RESERVA.CONFIRMADA || "CONFIRMADA";
    await reserva.save();

    // Notificar al cliente la confirmación
    const datosCorreo: DatosReservaCorreo = {
      to: reserva.cliente?.email,
      nombreCliente: reserva.cliente?.nombre || "Cliente",
      servicio: reserva.servicio?.nombre || "Servicio",
      profesional: reserva.profesional?.nombre,
      fecha: reserva.fecha,
      hora: reserva.hora,
      codigo: String(reserva._id).slice(-6).toUpperCase(),
    };

    MailService.enviarConfirmacionReserva(datosCorreo).catch((error) => {
      console.error("[MAIL ERROR] Error al enviar correo de confirmación:", error);
    });

    return reserva;
  }

  /**
   * Marca una reserva como completada.
   */
  static async completar(id: string, user: any) {
    const reserva: any = await ReservaRepository.buscarPorId(id);
    if (!reserva) {
      throw new Error("Reserva no encontrada");
    }

    reserva.estado = ESTADOS_RESERVA.COMPLETADA || "COMPLETADA";
    await reserva.save();
    return reserva;
  }

  /**
   * Cancela una reserva.
   */
  static async cancelar(id: string, user: any) {
    const reserva: any = await ReservaRepository.buscarPorId(id);
    if (!reserva) {
      throw new Error("Reserva no encontrada");
    }

    reserva.estado = ESTADOS_RESERVA.CANCELADA || "CANCELADA";
    await reserva.save();
    return reserva;
  }

  /**
   * Rechaza una reserva.
   */
  static async rechazar(id: string, user: any) {
    const reserva: any = await ReservaRepository.buscarPorId(id);
    if (!reserva) {
      throw new Error("Reserva no encontrada");
    }

    reserva.estado = ESTADOS_RESERVA.RECHAZADA || "RECHAZADA";
    await reserva.save();
    return reserva;
  }

  /**
   * Reprograma fecha y hora de la reserva, validando que no haya conflictos.
   */
  static async reprogramar(id: string, nuevaFecha: string, nuevaHora: string, user: any) {
    const reserva: any = await ReservaRepository.buscarPorId(id);
    if (!reserva) {
      throw new Error("Reserva no encontrada");
    }

    // Verificar si el nuevo horario está ocupado
    const existeConflicto = await ReservaRepository.existeConflicto(
      String(reserva.profesional._id || reserva.profesional),
      nuevaFecha,
      nuevaHora,
      id
    );

    if (existeConflicto) {
      throw new Error("El nuevo horario seleccionado no está disponible.");
    }

    reserva.fecha = nuevaFecha;
    reserva.hora = nuevaHora;
    await reserva.save();

    // Opcional: enviar mail de notificación por reprogramación
    const datosCorreo: DatosReservaCorreo = {
      to: reserva.cliente?.email,
      nombreCliente: reserva.cliente?.nombre || "Cliente",
      servicio: reserva.servicio?.nombre || "Servicio",
      profesional: reserva.profesional?.nombre,
      fecha: nuevaFecha,
      hora: nuevaHora,
      codigo: String(reserva._id).slice(-6).toUpperCase(),
    };

    MailService.enviarConfirmacionReserva(datosCorreo).catch((error) => {
      console.error("[MAIL ERROR] Error al enviar correo de reprogramación:", error);
    });

    return reserva;
  }

  /**
   * Consulta las horas disponibles de un profesional para un día específico.
   */
  static async disponibilidad(profesionalId: string, fecha: string) {
    const reservasOcupadas = await ReservaRepository.delDiaExcluyendoCanceladas(
      profesionalId,
      fecha
    );
    
    // Retorna la lista de reservas que ocupan espacio ese día
    return reservasOcupadas;
  }
}